diff --git a/src/runtime/client/remote-functions/form.svelte.js b/src/runtime/client/remote-functions/form.svelte.js
index e2b7fb0944fe079a273bc280de0ae17882da1ae4..b06739c452afcb61fad6e28a9c03be07b73c0c0b 100644
--- a/src/runtime/client/remote-functions/form.svelte.js
+++ b/src/runtime/client/remote-functions/form.svelte.js
@@ -376,6 +376,7 @@ export function form(id) {
 
 			/** @param {SubmitEvent} event */
 			const handle_submit = async (event) => {
+				if (event.defaultPrevented) return;
 				const form = /** @type {HTMLFormElement} */ (event.target);
 				const method = event.submitter?.hasAttribute('formmethod')
 					? /** @type {HTMLButtonElement | HTMLInputElement} */ (event.submitter).formMethod
@@ -533,6 +534,7 @@ export function form(id) {
 				// need to wait a moment, because the `reset` event occurs before
 				// the inputs are actually updated (so that it can be cancelled)
 				await tick();
+				await new Promise((resolve) => setTimeout(resolve, 0));
 
 				input = convert_formdata(new FormData(form));
 				raw_issues = [];
